pkg/fqdn/dnsproxy/proxy.go

```diff
@@ -268,52 +339,155 @@ func (p *DNSProxy) RemoveRestoredRules(endpointID uint16) {
 	p.removeRestoredRulesLocked(uint64(endpointID))
 }
 
+// lookupOrCompilePatternMatcher will check if the regex is already compiled and present in another policy, and
+// will reuse it in order to reduce memory consumption. The usage is reference counted, so all calls where
+// lookupOrCompilePatternMatcher returns no error, has to release it via releasePatternMatcher when its no longer being used
+// by the policy.
+func (allow perEPAllow) lookupOrCompilePatternMatcher(pattern string) (*regexp.Regexp, error) {
+	if entry, ok := allow.patternMatchersByPattern[pattern]; ok {
+		entry.referenceCount += 1
+		return entry.matcher, nil
+	}
+	regex, err := re.CompileRegex(pattern)
+	if err != nil {
+		return nil, err
+	}
+	allow.patternMatchersByPattern[pattern] = &patternMatcherEntry{matcher: regex, referenceCount: 1}
+	return regex, nil
+}
+
+// releasePatternMatcher releases the pattern matcher. In case there are no longer any references to it,
+// it will be freed.
+func (allow perEPAllow) releasePatternMatcher(patternMatcher *regexp.Regexp) {
+	if indexEntry, ok := allow.patternMatchersByPattern[patternMatcher.String()]; ok {
+		switch indexEntry.referenceCount {
+		case 1:
+			delete(allow.patternMatchersByPattern, patternMatcher.String())
+		default:
+			indexEntry.referenceCount -= 1
+		}
+	}
+}
+
+// releaseMatchers releases both the pattern and the name matchers for toe provided matcher
+func (allow perEPAllow) releaseMatchers(matcher fqdnMatcher) {
+	allow.releasePatternMatcher(matcher.patternMatcher)
+	allow.releaseNameMatcher(matcher.nameMatcher)
+}
+
+// lookupOrCreateNameMatcher will check if a map with the same list of matchNames is already compiled and present in another
+// policy, and will reuse it in order to reduce memory consumption. The usage is reference counted, so all calls to
+// lookupOrCreateNameMatcher, has to release it via releaseNameMatcher when its no longer being used
+// by the policy.
+func (allow perEPAllow) lookupOrCreateNameMatcher(fqdnList []string) *matchNameMatcher {
+	fqdnKey := strings.Join(fqdnList, "|")
+	if entry, ok := allow.nameMatcherBySignature[fqdnKey]; ok {
+		entry.referenceCount += 1
+		return entry.matcher
+	}
+	f := &matchNameMatcher{
+		mapping: make(map[string]struct{}, len(fqdnList)),
+		key:     &fqdnKey,
+	}
+	for _, fqdn := range fqdnList {
+		f.mapping[fqdn] = struct{}{}
+	}
+	allow.nameMatcherBySignature[fqdnKey] = &nameMatcherEntry{matcher: f, referenceCount: 1}
+	return f
+}
+
+// releaseNameMatcher releases the name matcher. In case there are no longer any references to
+// it, it will be freed.
```

Comment:
```suggestion
// releaseNameMatcher releases the entry from the name matcher map. In case there are no longer any references to
// it, it will be freed.
```

Just to clarify things + make it more explicit.

Reply:
Sounds good, thanks!